### PR TITLE
Fix maybe map_or/map_or_else return types.

### PR DESCRIPTION
**Pull request by jaheba**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/maybe.py [hunk 1]
```diff
@@ -75,9 +75,7 @@ def map(self: Optional[T], fn: Callable[[T], U]) -> Optional[U]:
     return map_or(self, None, fn)
 
 
-def map_or(
-    self: Optional[T], default: Optional[U], fn: Callable[[T], U]
-) -> Optional[U]:
+def map_or(self: Optional[T], default: Optional[U], fn: Callable[[T], U]) -> U:
```
Comment (lostella) [inline src/gluonts/maybe.py]: `default: U`?

**Comment by lostella** [comment]
Maybe you could also add docstrings? I find it a bit hard to read through what these do, maybe because of the ordering of the arguments (reading `map` I would expect a function first)

**Comment by lostella** [comment]
Also, did `mypy` not catch these?

File: src/gluonts/maybe.py (hunk 1 above)
Comment (jaheba) [reply src/gluonts/maybe.py]: Thanks!

**Comment by jaheba** [comment]
> Also, did `mypy` not catch these?

It does if you run it directly. But I think our current setup doesn't, because of the marker files.

**Comment by jaheba** [comment]
> Maybe you could also add docstrings? I find it a bit hard to read through what these do, maybe because of the ordering of the arguments (reading `map` I would expect a function first)

Granted, it is a bit weird. But I essentially followed Rust ordering of arguments, putting the `self` as the first one. I think the reason for putting the `fn` last is that you can then have longish inline functions, which are not easily doable with Python's lamdas.

**Review by lostella — APPROVED** [review]

**Pull request merged**